dpe/src/commands/derive_child.rs

```diff
@@ -159,7 +154,22 @@ impl CommandExecution for DeriveChildCmd {
             allow_x509: self.allows_x509(),
         });
 
-        dpe.add_tci_measurement(env, child_idx, &TciMeasurement(self.data), target_locality)?;
+        dpe.add_tci_measurement(
+            env,
+            &mut tmp_context,
+            &TciMeasurement(self.data),
+            target_locality,
+        )?;
+
+        if !self.retains_parent() {
+            dpe.contexts[parent_idx].state = ContextState::Retired;
+            dpe.contexts[parent_idx].handle = ContextHandle([0xff; ContextHandle::SIZE]);
+        } else if !dpe.contexts[parent_idx].handle.is_default() {
+            dpe.contexts[parent_idx].handle = dpe.generate_new_handle(env)?;
+        }
+
+        // at this point we cannot error out anymore, so it is safe to set the child context
+        dpe.contexts[child_idx] = tmp_context;
 
         // Add child to the parent's list of children.
         dpe.contexts[parent_idx].add_child(child_idx)?;
```

Comment:
If this fails, the parent_idx state is still modified. I think we need to make updating the parent `state`, `handle`, and `children` atomic.

Reply:
Refactored it to also copy the parent context into a temporary context. Also, added the same logic to sign and destroy_context to avoid mutation upon InternalErrors.